From: Jeff Moyer <jmoyer@redhat.com>
Date: Mon, 8 Sep 2008 13:48:52 -0400
Subject: [fs] check for device resize when rescanning partitions
Message-id: x497i9mv88r.fsf@segfault.boston.devel.redhat.com
O-Subject: [rhel5 patch 3/6 - v2] Check for device resize when rescanning partitions
Bugzilla: 444964
RH-Acked-by: Josef Bacik <jbacik@redhat.com>
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

Check for device resize when rescanning partitions

Check for device resize in the rescan_partitions() routine. If the device
has been resized, the bdev size is set to match. The rescan_partitions()
routine is called when opening the device and when calling the
BLKRRPART ioctl.

Signed-off-by: Andrew Patterson <andrew.patterson@hp.com>

diff --git a/fs/partitions/check.c b/fs/partitions/check.c
index d60d586..b8c4ab0 100644
--- a/fs/partitions/check.c
+++ b/fs/partitions/check.c
@@ -487,11 +487,12 @@ int rescan_partitions(struct gendisk *disk, struct block_device *bdev)
 	res = invalidate_partition(disk, 0);
 	if (res)
 		return res;
-	bdev->bd_invalidated = 0;
 	for (p = 1; p < disk->minors; p++)
 		delete_partition(disk, p);
 	if (disk->fops->revalidate_disk)
 		disk->fops->revalidate_disk(disk);
+	check_disk_size_change(disk, bdev);
+	bdev->bd_invalidated = 0;
 	if (!get_capacity(disk) || !(state = check_partition(disk, bdev)))
 		return 0;
 	if (IS_ERR(state))      /* I/O error reading the partition table */
